### PR TITLE
[Impeller] Update BlitPass::AddCopy to use destination_region instead of origin for buffer to texture copies.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Based on https://github.com/flutter/engine/pull/52510

Work towards https://github.com/flutter/flutter/issues/138798

Change IPoint destination_origin to IRect destination_region, which allows us to specify an area smaller than the entire texture to replace. This will eventually allow us to only upload individual glyphs. This fixes the cubemap issue I previously hit: each face needs to track initialization separately.

### CODE REVIEW

File: impeller/renderer/backend/gles/texture_gles.h [hunk 1]
```diff
@@ -60,13 +60,18 @@ class TextureGLES final : public Texture,
 
   std::optional<GLuint> GetFBO() const { return wrapped_fbo_; }
 
- private:
-  friend class AllocatorMTL;
+  // For non cubemap textures, 0 indicates uninitialized and 1 indicates
+  // initialized. For cubemap textures, each face is initialized separately with
+  // each bit tracking the initialization of the corresponding slice.
+  void MarkSliceInitialized(size_t slice) const;
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/texture_gles.h:66]: This is what I was missing last time. For the GLES cube map textures we need to tracking initialization per slice

File: impeller/renderer/backend/metal/blit_command_mtl.mm [hunk 2]
```diff
@@ -114,25 +114,22 @@
...
-      source_size_mtl.width * destination_bytes_per_pixel;
-  auto destination_bytes_per_image =
-      source_size_mtl.height * destination_bytes_per_row;
+      destination_region.GetWidth() * destination_bytes_per_pixel;
 
   [encoder copyFromBuffer:source_mtl
              sourceOffset:source.range.offset
         sourceBytesPerRow:destination_bytes_per_row
-      sourceBytesPerImage:destination_bytes_per_image
+      sourceBytesPerImage:0
```
Comment (jonahwilliams) [inline impeller/renderer/backend/metal/blit_command_mtl.mm:129]: according to https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400752-copyfrombuffer , we can set sourceBytesPerImage to 0 for 2d textures

**Comment by jonahwilliams** [comment]
Goldens look good:

image

File: impeller/aiks/aiks_unittests.cc [hunk 3]
```diff
@@ -3151,6 +3152,39 @@ TEST_P(AiksTest, CanRenderTextWithLargePerspectiveTransform) {
...
+  auto mapping =
+      std::make_shared<fml::NonOwnedMapping>(bytes.data(), bytes.size());
+  auto device_buffer =
+      GetContext()->GetResourceAllocator()->CreateBufferWithCopy(*mapping);
+  auto cmd_buffer = GetContext()->CreateCommandBuffer();
+  auto blit_pass = cmd_buffer->CreateBlitPass();
+  blit_pass->AddCopy(DeviceBuffer::AsBufferView(device_buffer), bridge,
+                     IRect::MakeLTRB(50, 50, 150, 150));
+
+  if (!blit_pass->EncodeCommands(GetContext()->GetResourceAllocator()) ||
```
Comment (chinmaygarde) [inline impeller/aiks/aiks_unittests.cc:3175]: Just curious. Why not ASSERT_TRUE? The GTEST_FAIL doesn't print out the assertion.

File: impeller/renderer/backend/gles/texture_gles.h [hunk 4]
```diff
@@ -60,13 +60,18 @@ class TextureGLES final : public Texture,
...
+
+  bool IsSliceInitialized(size_t slice) const;
 
  private:
   ReactorGLES::Ref reactor_;
   const Type type_;
   HandleGLES handle_;
-  mutable bool contents_initialized_ = false;
+  mutable uint32_t contents_initialized_ = 0;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/texture_gles.h:74]: Consider using [`std::bitset<6>`](https://en.cppreference.com/w/cpp/utility/bitset/bitset) and renaming to `slices_initialized_`. That gives you nice utilites to set and check these things.

File: impeller/renderer/backend/metal/blit_command_mtl.mm (hunk 2 above)
Comment (chinmaygarde) [reply impeller/renderer/backend/metal/blit_command_mtl.mm:129]: Let's add an inline comment to that effect? That just reads wrong lol.

File: impeller/renderer/backend/gles/blit_pass_gles.cc [hunk 5]
```diff
@@ -130,14 +130,14 @@ bool BlitPassGLES::OnCopyTextureToBufferCommand(
...
     std::string label,
     uint32_t slice) {
   auto command = std::make_unique<BlitCopyBufferToTextureCommandGLES>();
   command->label = label;
   command->source = std::move(source);
   command->destination = std::move(destination);
-  command->destination_origin = destination_origin;
+  command->destination_region = destination_region;
   command->label = label;
   command->slice = slice;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/blit_pass_gles.cc:142]: Here and elsewhere, we should check that slices are from 0 to 5. And, if not, either return false or clamp to the values. I prefer the former.

**Review by chinmaygarde — APPROVED** [review]
Minor nits and suggestions.

File: impeller/renderer/backend/gles/blit_pass_gles.cc (hunk 5 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/blit_pass_gles.cc:142]: I added this check it blit_pass.cc and a test case.

File: impeller/aiks/aiks_unittests.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/aiks/aiks_unittests.cc]: Fixed

File: impeller/renderer/backend/gles/texture_gles.h (hunk 4 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/texture_gles.h]: TIL

File: impeller/renderer/backend/metal/blit_command_mtl.mm (hunk 2 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/metal/blit_command_mtl.mm]: DONE

Comment (jonahwilliams) [reply impeller/renderer/backend/metal/blit_command_mtl.mm]: oops, I mean "Done"

**Pull request merged**